### PR TITLE
Add parallel downloads tests in cache_handler.go

### PR DESCRIPTION
**Pull request by sethiay**
### Description
Add parallel downloads tests in cache_handler.go

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - NA
3. Integration tests - NA

### CODE REVIEW

File: internal/cache/file/cache_handler_test.go [hunk 1]
```diff
@@ -62,457 +55,584 @@ type cacheHandlerTest struct {
...
+	fileInfoKeyName := addTestFileInfoEntryInCache(t, cache, object, storage.TestBucketName)
+	downloadPath := util.GetDownloadPath(cacheHandler.cacheDir, util.GetObjectPath(bucket.Name(), object.Name))
+	_, err = util.CreateFile(data.FileSpec{Path: downloadPath, FilePerm: util.DefaultFilePerm, DirPerm: util.DefaultDirPerm}, os.O_RDONLY)
+	require.NoError(t, err)
+
+	job := jobManager.CreateJobIfNotExists(object, bucket)
+	require.NotNil(t, job)
+
+	t.Cleanup(func() {
+		fakeStorage.ShutDown()
```
Comment (kislaykishore) [inline internal/cache/file/cache_handler_test.go:93]: The cleanup should happen right after fakeStorage was created. Otherwise, if the test fails in between e.g. the precondition "require.NotNil" in the previous line is not satisfied, then fakeStorage.Shutdown won't be called. Same goes for the cacheDir too.

File: internal/cache/file/cache_handler_test.go [hunk 2]
```diff
@@ -523,192 +643,330 @@ func (chrT *cacheHandlerTest) Test_GetCacheHandle_ConcurrentDifferentFiles() {
...
+				assert.Equal(t, false, doesFileExist(t, chTestArgs.downloadPath))
+				// File info should also be removed.
+				assert.False(t, isEntryInFileInfoCache(t, chTestArgs.cache, chTestArgs.object.Name, chTestArgs.bucket.Name()))
+			}
+
+			// Start concurrent GetCacheHandle()
+			for i := 0; i < 5; i++ {
+				wg.Add(1)
+				go InvalidateCacheTestFun(t)
+			}
```
Comment (kislaykishore) [inline internal/cache/file/cache_handler_test.go:789]: The function can access "t" from outer scope.

```suggestion
			InvalidateCacheTestFun := func() {
				defer wg.Done()

				err := chTestArgs.cacheHandler.InvalidateCache(chTestArgs.object.Name, chTestArgs.bucket.Name())

				assert.NoError(t, err)
				assert.NotNil(t, existingJob)
				assert.Equal(t, downloader.Invalid, existingJob.GetStatus().Name)
				assert.Equal(t, false, doesFileExist(t, chTestArgs.downloadPath))
				// File info should also be removed.
				assert.False(t, isEntryInFileInfoCache(t, chTestArgs.cache, chTestArgs.object.Name, chTestArgs.bucket.Name()))
			}

			// Start concurrent GetCacheHandle()
			for i := 0; i < 5; i++ {
				wg.Add(1)
				go InvalidateCacheTestFun()
			}
```

File: internal/cache/file/cache_handler_test.go [hunk 3]
```diff
@@ -523,192 +643,330 @@ func (chrT *cacheHandlerTest) Test_GetCacheHandle_ConcurrentDifferentFiles() {
...
-	for i := 0; i < 5; i++ {
-		wg.Add(1)
-		go InvalidateCacheTestFun(i)
+	for _, tc := range tbl {
+		t.Run(tc.name, func(t *testing.T) {
+			chTestArgs := initializeCacheHandlerTestArgs(t, &tc.fileCacheConfig, tc.cacheDir)
+			existingJob := getDownloadJobForTestObject(t, chTestArgs)
+			require.Equal(t, downloader.NotStarted, existingJob.GetStatus().Name)
+			require.True(t, isEntryInFileInfoCache(t, chTestArgs.cache, chTestArgs.object.Name, chTestArgs.bucket.Name()))
+			wg := sync.WaitGroup{}
```
Comment (kislaykishore) [inline internal/cache/file/cache_handler_test.go:771]: nit: sync.WaitGroup has a useful zero value i.e. `var wg sync.WaitGroup` is good enough. But this change is out of scope for this PR, so it's okay to ignore this.

**Review by kislaykishore — CHANGES_REQUESTED** [review]

File: internal/cache/file/cache_handler_test.go (hunk 1 above)
Comment (sethiay) [reply internal/cache/file/cache_handler_test.go]: Taken care in https://github.com/GoogleCloudPlatform/gcsfuse/pull/2178

File: internal/cache/file/cache_handler_test.go [hunk 4]
```diff
@@ -602,166 +683,293 @@ func Test_InvalidateCache_WhenEntryNotInCache(t *testing.T) {
...
+		expectedCacheHandleReadErr  error
+		expectedCInvalidateCacheErr error
+		expectedCacheFileReadErr    error
+	}{
+		{
+			name:                        "Non parallel downloads",
+			fileCacheConfig:             config.FileCacheConfig{EnableCRC: true},
+			cacheDir:                    path.Join(os.Getenv("HOME"), "CacheHandlerTest/dir"),
+			expectedCacheHandleReadErr:  nil,
+			expectedCInvalidateCacheErr: nil,
```
Comment (kislaykishore) [inline internal/cache/file/cache_handler_test.go:699]: Typo?
```suggestion
			expectedInvalidateCacheErr: nil,
```

File: internal/cache/file/cache_handler_test.go [hunk 5]
```diff
@@ -602,166 +683,293 @@ func Test_InvalidateCache_WhenEntryNotInCache(t *testing.T) {
...
+			cacheDir:               path.Join(os.Getenv("HOME"), "CacheHandlerTest/dir"),
+			expectedCacheHandleErr: nil,
+			expectedJobStatus:      []string{string(downloader.Completed)},
+		},
+		{
+			name: "Parallel downloads",
+			fileCacheConfig: config.FileCacheConfig{EnableCRC: true, EnableParallelDownloads: true,
+				ParallelDownloadsPerFile: 4, MaxParallelDownloads: 20, DownloadChunkSizeMB: 3},
+			cacheDir:               path.Join(os.Getenv("HOME"), "CacheHandlerTest/dir"),
+			expectedCacheHandleErr: fmt.Errorf("read via gcs"),
```
Comment (kislaykishore) [inline internal/cache/file/cache_handler_test.go:920]: Can you add a comment explaining why an error is returned for Parallel downloads?

File: internal/cache/file/cache_handler_test.go [hunk 6]
```diff
@@ -602,166 +683,293 @@ func Test_InvalidateCache_WhenEntryNotInCache(t *testing.T) {
...
-
-	// Start concurrent GetCacheHandle()
-	for i := 0; i < 5; i++ {
-		wg.Add(1)
-		go InvalidateCacheTestFun(i)
+	for _, tc := range tbl {
+		t.Run(tc.name, func(t *testing.T) {
+			chTestArgs := initializeCacheHandlerTestArgs(t, &tc.fileCacheConfig, tc.cacheDir)
+			wg := sync.WaitGroup{}
+			InvalidateCacheTestFun := func(index int) {
```
[the remark this answers is not in the document]
Comment (sethiay) [reply internal/cache/file/cache_handler_test.go]: yeah, it was already there. Corrected now.

File: internal/cache/file/cache_handler_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (sethiay) [reply internal/cache/file/cache_handler_test.go]: okay, modified for new tests, will do separately for all.

File: internal/cache/file/cache_handler_test.go [hunk 7]
```diff
@@ -899,26 +908,29 @@ func Test_InvalidateCache_GetCacheHandle_Concurrent(t *testing.T) {
...
 		{
 			name: "Parallel downloads",
 			fileCacheConfig: config.FileCacheConfig{EnableCRC: true, EnableParallelDownloads: true,
 				ParallelDownloadsPerFile: 4, MaxParallelDownloads: 20, DownloadChunkSizeMB: 3},
-			cacheDir:               path.Join(os.Getenv("HOME"), "CacheHandlerTest/dir"),
-			expectedCacheHandleErr: fmt.Errorf("read via gcs"),
-			expectedJobStatus:      []string{string(downloader.Completed), string(downloader.Invalid)},
+			cacheDir: path.Join(os.Getenv("HOME"), "CacheHandlerTest/dir"),
+			// Error is expected in parallel downloads because the foreground reads
+			// doesn't for async job to download till the requested offset unlike
```
Comment (kislaykishore) [inline internal/cache/file/cache_handler_test.go:930]: typo: 
```suggestion
			// doesn't wait for async job to download till the requested offset unlike
```

**Review by kislaykishore — APPROVED** [review]

**Review by kislaykishore — APPROVED** [review]

**Review by ashmeenkaur — APPROVED** [review]
Relying on @kislaykishore's review
